exporter/pulsarexporter/factory.go

```diff
@@ -0,0 +1,170 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pulsarexporter
+
+import (
+	"context"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+)
+
+const (
+	typeStr             = "pulsar"
+	defaultTracesTopic  = "otlp_spans"
+	defaultMetricsTopic = "otlp_metrics"
+	defaultLogsTopic    = "otlp_logs"
+	defaultEncoding     = "otlp_proto"
+	defaultBroker       = "pulsar://localhost:6650"
+)
+
+// FactoryOption applies changes to pulsarExporterFactory.
+type FactoryOption func(factory *pulsarExporterFactory)
+
+// WithTracesMarshalers adds tracesMarshalers.
+func WithTracesMarshalers(tracesMarshalers ...TracesMarshaler) FactoryOption {
+	return func(factory *pulsarExporterFactory) {
+		for _, marshaler := range tracesMarshalers {
+			factory.tracesMarshalers[marshaler.Encoding()] = marshaler
+		}
+	}
+}
+
+// NewFactory creates Pulsar exporter factory.
+func NewFactory(options ...FactoryOption) component.ExporterFactory {
+	f := &pulsarExporterFactory{
+		tracesMarshalers:  tracesMarshalers(),
+		metricsMarshalers: metricsMarshalers(),
+		logsMarshalers:    logsMarshalers(),
+	}
+	for _, o := range options {
+		o(f)
+	}
+	return component.NewExporterFactory(
+		typeStr,
+		createDefaultConfig,
+		component.WithTracesExporter(f.createTracesExporter),
+		component.WithMetricsExporter(f.createMetricsExporter),
+		component.WithLogsExporter(f.createLogsExporter),
+	)
+}
+
+func createDefaultConfig() config.Exporter {
+	return &Config{
+		ExporterSettings: config.NewExporterSettings(config.NewComponentID(typeStr)),
+		RetrySettings:    exporterhelper.NewDefaultRetrySettings(),
+		QueueSettings:    exporterhelper.NewDefaultQueueSettings(),
+		Endpoint:         defaultBroker,
+		// using an empty topic to track when it has not been set by user, default is based on traces or metrics.
+		Topic:          "",
+		Encoding:       defaultEncoding,
+		Authentication: Authentication{},
+	}
+}
+
+type pulsarExporterFactory struct {
+	tracesMarshalers  map[string]TracesMarshaler
+	metricsMarshalers map[string]MetricsMarshaler
+	logsMarshalers    map[string]LogsMarshaler
+}
+
+func (f *pulsarExporterFactory) createTracesExporter(
+	_ context.Context,
+	set component.ExporterCreateSettings,
+	cfg config.Exporter,
+) (component.TracesExporter, error) {
+	oCfg := cfg.(*Config)
+	if oCfg.Topic == "" {
+		oCfg.Topic = defaultTracesTopic
```

Comment:
This will screw up the configuration for other types exporters, e.g.  `otlp_spans` can be set for metric or/and logs. Please don't mutate the original config

Reply:
refers to https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/kafkaexporter/factory.go , do I need to remove it?

Reply:
I believe not, I debugged it, each component receives it's own config. I mean, the config only passed to its own component, won't passed to other types components.

Reply:
![QQ20220709-0](https://user-images.githubusercontent.com/95597048/178049055-6cbfb40a-bcab-4d7b-9c73-9716afd5d211.png)
![QQ20220709-1](https://user-images.githubusercontent.com/95597048/178049068-8e8b464d-7529-4c03-bb04-0c250851d217.png)
As you can see, config will passed to its own component, it won't screw up other types exporters. but, fine, I'll change it if you insist

Reply:
The screenshots don't show pointer addresses. Not sure what you wanted to prove

Reply:
```
	oCfg := *(cfg.(*Config))
	if oCfg.Topic == "" {
		oCfg.Topic = defaultMetricsTopic
	}
```
I copied the config instance, so that it won't effect other types exporters